### PR TITLE
Stock symbol not exist for historic endpoint

### PR DESCRIPTION
**Pull request by murradkh**

### CODE REVIEW

File: myapp/stock_api.py [hunk 1]
```diff
@@ -57,4 +57,12 @@ def get_stock_info(symbol):
...
-    return _request_data('/stable/stock/{symbol}/chart/{time_range}'.format(symbol=symbol, time_range=time_range))
+    try:
+        return _request_data('/stable/stock/{symbol}/chart/{time_range}'.format(symbol=symbol, time_range=time_range))
+    except ConnectionError:
+        raise StockServerUnReachable("Stock server UnReachable!")
+    except Exception as e:
+        for arg in e.args:
+            if isinstance(arg, dict) and b"Unknown symbol" in arg.values():
+                raise StockSymbolNotFound("Unknown Stock Symbol!")
+        raise e
```
Comment (kamaql97) [inline myapp/stock_api.py:68]: Just a question: if the condition defined on line 66 is true, would two expections be raised?

File: myapp/stock_api.py [hunk 2]
```diff
@@ -57,4 +57,12 @@ def get_stock_info(symbol):
...
 def get_stock_historic_prices(symbol, time_range='1m'):
-    return _request_data('/stable/stock/{symbol}/chart/{time_range}'.format(symbol=symbol, time_range=time_range))
+    try:
+        return _request_data('/stable/stock/{symbol}/chart/{time_range}'.format(symbol=symbol, time_range=time_range))
+    except ConnectionError:
+        raise StockServerUnReachable("Stock server UnReachable!")
+    except Exception as e:
+        for arg in e.args:
+            if isinstance(arg, dict) and b"Unknown symbol" in arg.values():
+                raise StockSymbolNotFound("Unknown Stock Symbol!")
```
Comment (kamalq97) [inline myapp/stock_api.py:67]: I need to remember to add this to my `watchlist_add` and `watchlist_remove` views in pull request #6 👍

File: myapp/views.py [hunk 3]
```diff
@@ -1,51 +1,83 @@
...
+# API for a stock's price over time
+# symbol is the requested stock's symbol ('AAPL' for Apple)
+# The response is JSON data of an array composed of "snapshot" objects (date + stock info + ...), usually one per day
+def single_stock_historic(request, symbol):
+    context = None
+    status_code = 200
+    try:
+        data = stock_api.get_stock_historic_prices(symbol, time_range='1m')
+        context = {'data': data}
+    except StockSymbolNotFound as e:
```
Comment (kamalq97) [inline myapp/views.py:77]: Very well written and nicely handled 👍

File: myapp/views.py [hunk 4]
```diff
@@ -1,51 +1,83 @@
...
+        response = render(request, 'exception.html', {'404_error_message': e.message})
+        response.status_code = 404
+        return response
+    except StockServerUnReachable as e:
+        response = render(request, 'exception.html', {'503_error_message': e.message})
+        response.status_code = 503  # Service Unavailable code
+        return response
+    except Exception as e:
+        response = render(request, 'exception.html',
+                          {'404_error_message': "Unknown Error occurred: {}".format(", ".join(e.args))})
```
Comment (kamalq97) [inline myapp/views.py]: In your context dictionary, change the error message key string to something other than `'404_error_message'`

**Review by kamalq97 — COMMENTED** [review]
Overall, this looks good, only one very minor edit is needed in the **views.py** file

File: myapp/stock_api.py (hunk 2 above)
Comment (murradkh) [reply myapp/stock_api.py:67]: yes, if you feel we need to centralize the exception handling in one place would be fine also.

File: myapp/views.py (hunk 4 above)
Comment (murradkh) [reply myapp/views.py]: yes i already changed it in the branch `non-existend-symbol`, this part of the code doesn't belong to here. mistakenly merged changes from there to this branch, I will remove the merges and make it separate

Comment (murradkh) [reply myapp/views.py]: mmm, looks like there is some changes in `non-existent-symbol` needed here, it's really a problem to have branches counting on other branches didn't merge, causing little confusion. 
whatever, I merged the updated changes from there to this branch, now should be fine

File: myapp/views.py [hunk 5]
```diff
@@ -17,23 +17,29 @@ def index(request):
...
-        response.status_code = 520  # Unknown Error
-        return response
+        status_code = 520  # Unknown Error
+        context = {'error_message': "Unknown Error occurred: {}".format(", ".join(e.args)), "status_code": status_code}
+        template = "exception.html"
     else:
-        return render(request, 'single_stock.html', {'page_title': 'Stock Page - %s' % symbol, 'data': data})
+        context = {'page_title': 'Stock Page - %s' % symbol, 'data': data}
+    finally:
+        resp = render(request, template, context)
```
Comment (kamalq97) [inline myapp/views.py]: Please call it `response` because I have OCD

**Review by kamalq97 — COMMENTED** [review]
The code seems fine, I will need to run it locally and check if there are no issues/bugs

**Review by kamalq97 — APPROVED** [review]
All seems good

**Pull request merged**